### DISCUSSION TITLE
FromRequest for Option<MyType>

### DISCUSSION THREAD

**Discussion by shotor** [actix/actix-web#2154] [Q&A] [answered]
I'm trying to implement a cookie parser using the `FromRequest` trait on a custom `Cookies` struct. It works fine but I would like to implement the trait for `Option<Cookies>` instead of `Cookies`. But as soon as I do this I get the error:

```
11 | impl FromRequest for Option<Cookies> {
   | ^^^^^^^^^^^^^^^^^^^^^---------------
   | |                    |
   | |                    `Option` is not defined in the current crate
   | impl doesn't use only types from inside the current crate
```

If I leave out Option it works fine. But I want 3 states: `Ok` (has cookies and valid), `None` (doesn't have cookies at all), `Error` (something went wrong). 

```rust
use actix_web::{dev, Error, FromRequest, HttpRequest};
use futures_util::future::{ok, Ready};
use std::option::Option;

#[derive(Debug)]
pub struct Cookies {
  pub session: String,
  pub refresh: String,
}

impl FromRequest for Cookies { // works
// impl FromRequest for Option<Cookies> { // doesn't work
  type Error = Error;
  type Future = Ready<Result<Self, Self::Error>>;
  type Config = ();

  fn from_request(
    req: &HttpRequest,
    payload: &mut dev::Payload,
  ) -> Self::Future {
    ...
  }
}
```

Any advice how to do this?

**Comment by robjtede** [comment, marked as answer]
It's an overlapping impl. Actix Web already implements `FromRequest for Option<T> where T: FromRequest`

**Reply by shotor** [reply]
Ok, so then I do this to try it:

```rust
  fn from_request(
    req: &HttpRequest,
    payload: &mut dev::Payload,
  ) -> Self::Future {
    ok(None.into())
    // or ok(None)
}
```

But then I get another error:

```
error[E0277]: the trait bound `Cookies: From<Option<_>>` is not satisfied
  --> server/src/context.rs:22:8
   |
22 |     ok(None.into())
   |        ^^^^^^^^^^^ the trait `From<Option<_>>` is not implemented for `Cookies`

// or for ok(None)

error[E0308]: mismatched types
  --> server/src/context.rs:22:8
   |
22 |     ok(None)
   |        ^^^^ expected struct `Cookies`, found enum `Option`
   |
   = note: expected struct `Cookies`
                found enum `Option<_>`
```

**Reply by robjtede** [reply]
If you return an error from your FromRequest impl, then using `Option<T>` will turn an `Err(_)` into `None`.

**Reply by robjtede** [reply]
https://docs.rs/actix-web/3.3.2/src/actix_web/extract.rs.html#93-114

**Reply by shotor** [reply]
I see, changing the type parameter on the route handler runs the other trait.

And there's no implementation for the trait

`impl<T> FromRequest for Result<Option<T>, T::Error>`

So there's no way to do this? Unless I define and implement the trait myself.

```rust
/// extract `Thing` from request
async fn index(supplied_thing: Result<Option<Thing>, Err>) -> String {
     match supplied_thing {
         Ok(thing) => match thing {
           Some(x) => ...,
           None => ....,
        },
         Err(e) => format!("Error extracting thing: {}", e)
    }
}
```

**Reply by robjtede** [reply]
Does using `Result<Option<...>, ...>` not work?

Alternatively, you could set up your own Option like enum and impl FromRequest for it.

**Reply by shotor** [reply]
The thing is I can't get it to compile, this doesn't work:

```rust
impl FromRequest for Cookies {
  type Error = Error;
  type Future = Ready<Result<Self, Self::Error>>;
  type Config = ();

  fn from_request(
    req: &HttpRequest,
    payload: &mut dev::Payload,
  ) -> Self::Future {
    ok(None) // errors here
  }
```

```
error[E0308]: mismatched types
  --> server/src/context.rs:22:8
   |
22 |     ok(None)
   |        ^^^^ expected struct `Cookies`, found enum `Option`
   |
   = note: expected struct `Cookies`
                found enum `Option<_>`
```

**Reply by fakeshadow** [reply]
Of course it doesn't work. `FromRequest::Future` must be a future output `Self` in `Result::Ok`.

What you want to do is something like this.

```rust
#[get("/")]
async fn index(foo: Result<Option<Foo>, Error>) -> &'static str {
    "Hello World!"
}

struct Foo;

impl FromRequest for Foo {
    type Error = Error;
    type Future = Ready<Result<Self, Self::Error>>;
    type Config = ();
  
    fn from_request(
      req: &HttpRequest,
      payload: &mut Payload,
    ) -> Self::Future {
      ready(Ok(Foo))
    }
}
```

When you impl a trait for some type. The associate types constriant must be respected and can not be override.

**Reply by shotor** [reply]
In your example the `foo` argument in index will either be `Foo` or `Error` (simply because it won't compile with a None result). I'm trying to get it to be `None` | `Error` | `Foo`. But in your example `ready(ok(None))` is also not allowed. Because as you say:

> When you impl a trait for some type. The associate types constriant must be respected and can not be override.

Which I don't fully understand, because these are both defined in the actix source:

```rust
impl<T: 'static> FromRequest for Option<T> { }
impl<T> FromRequest for Result<T, T::Error> { }
```

My guess is I need something like:

```rust
impl<T> FromRequest for Result<Option<T>, T::Error>
```

**Reply by fakeshadow** [reply]
I'm just explaining how trait impl work. You error is because your trait want `Result<T, E>` as output when you actually trying to return `Result<Option<T>, E>`. 

`impl<T: FromRequest> FromRequest for Result<T, T::Error> { }`
Notice `T` is bound to `FromRequest` trait so it include `Option<T as FromRequest>`.

**Reply by shotor** [reply]
> I'm just explaining how trait impl work. You error is because your trait want Result<T, E> as output when you actually trying to return Result<Option<T>, E>.

This part is clear

> impl<T: FromRequest> FromRequest for Result<T, T::Error> { }
Notice T is bound to FromRequest trait so it include Option<T as FromRequest>.

This is still confusing. Need to look at simpler examples.

In any case, I'm skipping this and wrapping the `None` case in an error. And handling it where I need the actual cookies.

**Reply by fakeshadow** [reply]
In fact if you want you can extract something like this. It's not like it has any meaning but it explains how the default impl works.

```rust
#[get("/")]
async fn index(foo: Result<Result<Option<Option<HttpRequest>>, Error>, Error>) -> &'static str {
    "Hello World!"
}
```

**Reply by shotor** [reply]
I'm guessing that's because you can substitute the generic `T` in the trait implementations with `Option<T>` and `Result<T, T::Error>` indefinitely.

`Result<T, T::Error>`
`Result<Option<T>, T::Error>`
`Result<Result<Option<T>, T::Error>, T::Error>>`
...etc

So the compiler will happily apply them. Even the one I want: `Result<Option<context::Cookies>, actix_web::Error>,`

But when implementing `FromRequest` to actually return a result option you're blocked because there's no implementation for `Result<Option, E>`.

**Reply by fakeshadow** [reply]
That's the rule of Rust. You can look up "orphan rules".
